File: CabClient.java
// ===== CLIENT APPLICATION =====
// CabClient.java
import java.rmi.Naming;
import java.util.List;
import java.util.Scanner;

public class CabClient {
    private static CabService cabService;
    private static Scanner scanner;
    private static String currentUser;
    
    public static void main(String[] args) {
        try {
            // Connect to RMI service
            cabService = (CabService) Naming.lookup("//localhost/CabService");
            scanner = new Scanner(System.in);
            
            System.out.println("=== CAB MANAGEMENT CLIENT ===");
            System.out.println("Connected to server successfully!");
            
            // Main menu loop
            while (true) {
                if (currentUser == null) {
                    showLoginMenu();
                } else {
                    showMainMenu();
                }
            }
            
        } catch (Exception e) {
            System.err.println("Client error: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    private static void showLoginMenu() {
        try {
            System.out.println("\n=== LOGIN MENU ===");
            System.out.println("1. Register");
            System.out.println("2. Login");
            System.out.println("3. Exit");
            System.out.print("Choose option: ");
            
            int choice = scanner.nextInt();
            scanner.nextLine(); // consume newline
            
            switch (choice) {
                case 1:
                    registerUser();
                    break;
                case 2:
                    loginUser();
                    break;
                case 3:
                    System.out.println("Goodbye!");
                    System.exit(0);
                    break;
                default:
                    System.out.println("Invalid choice!");
            }
        } catch (Exception e) {
            System.err.println("Error in login menu: " + e.getMessage());
            scanner.nextLine(); // clear invalid input
        }
    }
    
    private static void showMainMenu() {
        try {
            System.out.println("\n=== MAIN MENU === (User: " + currentUser + ")");
            System.out.println("1. Book a Cab");
            System.out.println("2. Check Ride Status");
            System.out.println("3. Cancel Ride");
            System.out.println("4. View Available Cabs");
            System.out.println("5. Set Driver Available (Driver only)");
            System.out.println("6. System Stats");
            System.out.println("7. Logout");
            System.out.print("Choose option: ");
            
            int choice = scanner.nextInt();
            scanner.nextLine(); // consume newline
            
            switch (choice) {
                case 1:
                    bookCab();
                    break;
                case 2:
                    checkRideStatus();
                    break;
                case 3:
                    cancelRide();
                    break;
                case 4:
                    viewAvailableCabs();
                    break;
                case 5:
                    setDriverAvailable();
                    break;
                case 6:
                    showSystemStats();
                    break;
                case 7:
                    currentUser = null;
                    System.out.println("Logged out successfully!");
                    break;
                default:
                    System.out.println("Invalid choice!");
            }
        } catch (Exception e) {
            System.err.println("Error in main menu: " + e.getMessage());
            scanner.nextLine(); // clear invalid input
        }
    }
    
    private static void registerUser() {
        try {
            System.out.print("Enter username: ");
            String username = scanner.nextLine();
            System.out.print("Enter password: ");
            String password = scanner.nextLine();
            System.out.print("User type (RIDER/DRIVER): ");
            String userType = scanner.nextLine().toUpperCase();
            
            boolean success = cabService.registerUser(username, password, userType);
            if (success) {
                System.out.println("Registration successful!");
            } else {
                System.out.println("Registration failed - user already exists!");
            }
        } catch (Exception e) {
            System.err.println("Registration error: " + e.getMessage());
        }
    }
    
    private static void loginUser() {
        try {
            System.out.print("Enter username: ");
            String username = scanner.nextLine();
            System.out.print("Enter password: ");
            String password = scanner.nextLine();
            
            boolean success = cabService.authenticateUser(username, password);
            if (success) {
                currentUser = username;
                System.out.println("Login successful! Welcome " + username);
            } else {
                System.out.println("Invalid credentials!");
            }
        } catch (Exception e) {
            System.err.println("Login error: " + e.getMessage());
        }
    }
    
    private static void bookCab() {
        try {
            System.out.print("Enter pickup location: ");
            String pickup = scanner.nextLine();
            System.out.print("Enter destination: ");
            String destination = scanner.nextLine();
            
            String result = cabService.bookCab(currentUser, pickup, destination);
            if ("NO_DRIVER_AVAILABLE".equals(result)) {
                System.out.println("Sorry, no drivers available at the moment!");
            } else {
                System.out.println("Cab booked successfully! Ride ID: " + result);
            }
        } catch (Exception e) {
            System.err.println("Booking error: " + e.getMessage());
        }
    }
    
    private static void checkRideStatus() {
        try {
            System.out.print("Enter ride ID: ");
            String rideId = scanner.nextLine();
            
            String status = cabService.getRideStatus(rideId);
            System.out.println("Status: " + status);
        } catch (Exception e) {
            System.err.println("Status check error: " + e.getMessage());
        }
    }
    
    private static void cancelRide() {
        try {
            System.out.print("Enter ride ID to cancel: ");
            String rideId = scanner.nextLine();
            
            boolean success = cabService.cancelRide(rideId);
            if (success) {
                System.out.println("Ride cancelled successfully!");
            } else {
                System.out.println("Could not cancel ride - check ride ID or status!");
            }
        } catch (Exception e) {
            System.err.println("Cancellation error: " + e.getMessage());
        }
    }
    
    private static void viewAvailableCabs() {
        try {
            System.out.print("Enter your location: ");
            String location = scanner.nextLine();
            
            List<String> availableCabs = cabService.getAvailableCabs(location);
            System.out.println("\nAvailable Cabs:");
            if (availableCabs.isEmpty()) {
                System.out.println("No cabs available in your area.");
            } else {
                for (String cab : availableCabs) {
                    System.out.println("- " + cab);
                }
            }
        } catch (Exception e) {
            System.err.println("Error fetching available cabs: " + e.getMessage());
        }
    }
    
    private static void setDriverAvailable() {
        try {
            System.out.print("Enter your current location: ");
            String location = scanner.nextLine();
            
            boolean success = cabService.setDriverAvailable(currentUser, location);
            if (success) {
                System.out.println("You are now available for rides!");
            } else {
                System.out.println("Error: Only registered drivers can set availability!");
            }
        } catch (Exception e) {
            System.err.println("Error setting availability: " + e.getMessage());
        }
    }
    
    private static void showSystemStats() {
        try {
            int activeRides = cabService.getActiveRides();
            int availableDrivers = cabService.getAvailableDrivers();
            
            System.out.println("\n=== SYSTEM STATISTICS ===");
            System.out.println("Active Rides: " + activeRides);
            System.out.println("Available Drivers: " + availableDrivers);
        } catch (Exception e) {
            System.err.println("Error fetching system stats: " + e.getMessage());
        }
    }
}

File: WebSocket/cab_client.py
import asyncio
import websockets
import json
import sys

class CabClient:
    def __init__(self, server_uri="ws://localhost:8765"):
        self.server_uri = server_uri
        self.websocket = None
        self.current_user = None
        self.user_type = None
        print(f"=== CAB MANAGEMENT WEBSOCKET CLIENT ===")
    
    async def connect(self):
        """Connect to WebSocket server"""
        try:
            self.websocket = await websockets.connect(self.server_uri)
            print(f"Connected to Cab Service at {self.server_uri}")
            return True
        except Exception as e:
            print(f"Failed to connect to server: {e}")
            return False
    
    async def send_request(self, action, **kwargs):
        """Send request to server and get response"""
        try:
            request = {"action": action, **kwargs}
            await self.websocket.send(json.dumps(request))
            
            response = await self.websocket.recv()
            return json.loads(response)
        except Exception as e:
            print(f"Error communicating with server: {e}")
            return {"success": False, "message": "Communication error"}
    
    def show_login_menu(self):
        """Display login menu"""
        print("\n=== LOGIN MENU ===")
        print("1. Register")
        print("2. Login")
        print("3. Exit")
        print("-" * 30)
    
    def show_main_menu(self):
        """Display main menu based on user type"""
        print(f"\n=== MAIN MENU === (User: {self.current_user})")
        print("1. Book a Cab")
        print("2. Check Ride Status")
        print("3. Cancel Ride")
        print("4. View Available Cabs")
        if self.user_type == "DRIVER":
            print("5. Set Driver Available (Driver only)")
        print("6. System Stats")
        print("7. Logout")
        print("-" * 50)
    
    async def register_user(self):
        """Register a new user"""
        try:
            print("\n--- Register New User ---")
            username = input("Enter username: ").strip()
            password = input("Enter password: ").strip()
            user_type = input("User type (RIDER/DRIVER): ").strip().upper()
            
            if not username or not password or user_type not in ["RIDER", "DRIVER"]:
                print("Invalid input! Please provide valid username, password, and user type.")
                return
            
            result = await self.send_request("register_user", 
                                           username=username, 
                                           password=password, 
                                           user_type=user_type)
            
            if result["success"]:
                print("Registration successful!")
            else:
                print(f"Registration failed: {result['message']}")
                
        except Exception as e:
            print(f"Error during registration: {e}")
    
    async def login_user(self):
        """Login user"""
        try:
            print("\n--- Login ---")
            username = input("Enter username: ").strip()
            password = input("Enter password: ").strip()
            
            if not username or not password:
                print("Username and password are required!")
                return
            
            result = await self.send_request("authenticate_user",
                                           username=username,
                                           password=password)
            
            if result["success"]:
                self.current_user = username
                self.user_type = result["user_type"]
                print(f"Login successful! Welcome {username} ({self.user_type})")
            else:
                print(f"Login failed: {result['message']}")
                
        except Exception as e:
            print(f"Error during login: {e}")
    
    async def book_cab(self):
        """Book a cab"""
        try:
            print("\n--- Book a Cab ---")
            pickup = input("Enter pickup location: ").strip()
            destination = input("Enter destination: ").strip()
            
            if not pickup or not destination:
                print("Both pickup and destination are required!")
                return
            
            result = await self.send_request("book_cab",
                                           username=self.current_user,
                                           pickup=pickup,
                                           destination=destination)
            
            if result["success"]:
                print(f"Cab booked successfully!")
                print(f"Ride ID: {result['ride_id']}")
                print(f"Driver: {result['driver']}")
                print(f"Fare: ₹{result['fare']}")
            else:
                print(f"Booking failed: {result['message']}")
                
        except Exception as e:
            print(f"Error booking cab: {e}")
    
    async def check_ride_status(self):
        """Check ride status"""
        try:
            print("\n--- Check Ride Status ---")
            ride_id = input("Enter ride ID: ").strip()
            
            if not ride_id:
                print("Ride ID is required!")
                return
            
            result = await self.send_request("get_ride_status", ride_id=ride_id)
            
            if result["success"]:
                print(f"Status: {result['ride_info']}")
            else:
                print(f"Error: {result['message']}")
                
        except Exception as e:
            print(f"Error checking ride status: {e}")
    
    async def cancel_ride(self):
        """Cancel a ride"""
        try:
            print("\n--- Cancel Ride ---")
            ride_id = input("Enter ride ID to cancel: ").strip()
            
            if not ride_id:
                print("Ride ID is required!")
                return
            
            result = await self.send_request("cancel_ride", ride_id=ride_id)
            
            if result["success"]:
                print(result["message"])
            else:
                print(f"Cancellation failed: {result['message']}")
                
        except Exception as e:
            print(f"Error cancelling ride: {e}")
    
    async def view_available_cabs(self):
        """View available cabs"""
        try:
            print("\n--- Available Cabs ---")
            location = input("Enter your location: ").strip()
            
            if not location:
                print("Location is required!")
                return
            
            result = await self.send_request("get_available_cabs", location=location)
            
            if result["success"]:
                cabs = result["cabs"]
                if cabs:
                    print(f"\nAvailable Cabs ({result['count']}):")
                    for cab in cabs:
                        print(f"- {cab}")
                else:
                    print("No cabs available in your area.")
            else:
                print("Error fetching available cabs")
                
        except Exception as e:
            print(f"Error fetching available cabs: {e}")
    
    async def set_driver_available(self):
        """Set driver as available"""
        try:
            print("\n--- Set Driver Available ---")
            location = input("Enter your current location: ").strip()
            
            if not location:
                print("Location is required!")
                return
            
            result = await self.send_request("set_driver_available",
                                           driver_name=self.current_user,
                                           location=location)
            
            if result["success"]:
                print(result["message"])
            else:
                print(f"Error: {result['message']}")
                
        except Exception as e:
            print(f"Error setting driver availability: {e}")
    
    async def show_system_stats(self):
        """Show system statistics"""
        try:
            print("\n--- System Statistics ---")
            
            active_rides_result = await self.send_request("get_active_rides")
            available_drivers_result = await self.send_request("get_available_drivers")
            
            if active_rides_result["success"] and available_drivers_result["success"]:
                print(f"Active Rides: {active_rides_result['count']}")
                print(f"Available Drivers: {available_drivers_result['count']}")
            else:
                print("Error fetching system statistics")
                
        except Exception as e:
            print(f"Error fetching system stats: {e}")
    
    async def run(self):
        """Main client loop"""
        if not await self.connect():
            return
        
        try:
            while True:
                if not self.current_user:
                    self.show_login_menu()
                    choice = input("Choose option (1-3): ").strip()
                    
                    if choice == '1':
                        await self.register_user()
                    elif choice == '2':
                        await self.login_user()
                    elif choice == '3':
                        print("Goodbye!")
                        break
                    else:
                        print("Invalid choice!")
                else:
                    self.show_main_menu()
                    max_choice = 7 if self.user_type == "RIDER" else 7
                    choice = input(f"Choose option (1-{max_choice}): ").strip()
                    
                    if choice == '1':
                        await self.book_cab()
                    elif choice == '2':
                        await self.check_ride_status()
                    elif choice == '3':
                        await self.cancel_ride()
                    elif choice == '4':
                        await self.view_available_cabs()
                    elif choice == '5' and self.user_type == "DRIVER":
                        await self.set_driver_available()
                    elif choice == '6':
                        await self.show_system_stats()
                    elif choice == '7':
                        self.current_user = None
                        self.user_type = None
                        print("Logged out successfully!")
                    else:
                        print("Invalid choice!")
                        
        except KeyboardInterrupt:
            print("\nGoodbye!")
        except Exception as e:
            print(f"Unexpected error: {e}")
        finally:
            if self.websocket:
                await self.websocket.close()

async def main():
    client = CabClient()
    await client.run()

if __name__ == "__main__":
    asyncio.run(main())